### ISSUE TITLE
Port remaining parallelizable aggregate functions to off-heap data structures

### ISSUE THREAD

**Issue by puzpuzpuz** [questdb/questdb#4120]
### Is your feature request related to a problem?

#4097 ported `min(str)`, `max(str)`, as well as `count_distinct()` for long, int, and IPv4 types to parallel GROUP BY, but some functions remain unported. Namely:
- [x] `count_distinct(uuid)`: requires a new long128 hash set, similar to the `GroupByLongHashSet` one
- [x] `count_distinct(long256)`: requires a new long256 hash set, similar to the `GroupByLongHashSet` one
- [ ] `approx_percentile(double)`: this one is tricky as we'll have to port HdrHistogram to become off-heap and flyweight
- [x] all `first`/`last` and `first_not_null`/`last_not_null` functions: to port them, we'll have to access and store row ids in the group by map
- [ ] `isOrdered(IPv4)`/`isOrdered(long)` functions: again, we need to track row ids
- [x] `ksum`/`nsum`

There is also `count_distinct(symbol)`, but we have early exit logic in that function (see #3974), so we don't want to port it, at least for now.

### Full Name:

Andrei Pechkurov

### Affiliation:

QuestDB

**Comment by puzpuzpuz** [comment]
@nwoolmer that's awesome, thank you! Look forward to your contribution.

**Comment by mcadariu** [comment]
Interested in picking up string_agg(str), if still available!

**Comment by puzpuzpuz** [comment]
Both `string_agg` and `string_distinct_agg` are still available. Yet, please keep in mind that these two are tricky as they require writing sophisticated off-heap data structs such as string hash sets.

**Comment by mcadariu** [comment]
Thanks @puzpuzpuz! 
In this PR I worked on `string_agg`: https://github.com/questdb/questdb/pull/6218. 
Looking forward to your feedback!

**Comment by mcadariu** [comment]
In PR https://github.com/questdb/questdb/pull/6237 I took a shot at adding `GroupByArraySink`. This is a precursor step for porting `last(array)`. Thanks!

**Comment by mcadariu** [comment]
In https://github.com/questdb/questdb/pull/6291 I've implemented `last(array)`. Note that it is dependent on https://github.com/questdb/questdb/pull/6237.

**Comment by mcadariu** [comment]
@puzpuzpuz the refactoring of `first(array)` is ready: https://github.com/questdb/questdb/pull/6322

**Comment by mcadariu** [comment]
This PR adds `first_not_null(array)`: https://github.com/questdb/questdb/pull/6344

**Comment by mcadariu** [comment]
Now I got the hang of it`last_not_null(array)` is ready: https://github.com/questdb/questdb/pull/6368 😄

**Comment by puzpuzpuz** [comment]
That's awesome! I'll take a look once we merge the prior PR, so that it's easier to see the diff. The PR should be merged soon.

**Comment by mcadariu** [comment]
@puzpuzpuz For the `isOrdered(IPv4)` is it maybe the same story as [this](https://github.com/questdb/questdb/pull/6218#issuecomment-3369276341)?

**Comment by puzpuzpuz** [comment]
> [@puzpuzpuz](https://github.com/puzpuzpuz) For the `isOrdered(IPv4)` is it maybe the same story as [this](https://github.com/questdb/questdb/pull/6218#issuecomment-3369276341)?

Yeah, this one is tough to make parallel because of having to handle gaps in the computed data to be able to merge partial results. Also, that function is not very popular and no one ever asked for a parallel version of it.

**Comment by mcadariu** [comment]
For the `approx_percentile(double)`, seems like it should be doable like the other ones. Create a histogram sink with the counts and other metadata, store the pointer, merge them (addition), get percentile.

**Comment by puzpuzpuz** [comment]
The problem with `approx_percentile()` is that we're using `HdrHistogram` library under the hood. So, to make it parallel, we need to change the library to off-heap data and implement merge function to combine multiple histograms into a single one.

**Comment by mcadariu** [comment]
Right, it could be used then as a reference implementation to compare against for ensuring correctness.

**Comment by mcadariu** [comment]
Histogram PR is ready:  https://github.com/questdb/questdb/pull/6502
